### PR TITLE
Add test for releasing a shared EVP_PKEY across threads and fix memory ordering for refcounts

### PR DESCRIPTION
**Pull request by t8m**
Add a test for #25660 and fix the memory ordering as proposed there.

### CODE REVIEW

**Comment by slontis** [comment]
@nhorman..

File: test/threadstest.c
Comment (nhorman) [inline test/threadstest.c:1031]: So, shared_evp_key should have a refcount of 1.  Ok

File: test/threadstest.c
Comment (nhorman) [inline test/threadstest.c:1034]: Then we iterate from I = 1 to 9 here, so at the end of this refcount should be 10 good, sorry, just ha to write that down to show myself the counting was right

**Review by nhorman — APPROVED** [review]
Ack, looks good.  We may also want to create a fuzzer that just hammers the refcount code.  I haven't thought through it all, but I'm not convinced that we're not susceptible to a near zero race using the relaxed memory ordering, particularly on weak memory ordered systems.

I'm specifically wondering about the case where

* CPU 0 drops refcount a to 0
* CPU 1 ups the refcount, seeing zero so it "does the right thing"
* CPU 2  ups the refcount,seeing the Internet from CPU 2, and treats the object as live when in fact it's being freed by CPU 1

I might be wrong and we're likely often protected by higher layer locks any way, but it might be worth looking at further

**Comment by t8m** [comment]
> * CPU 0 drops refcount a to 0
> 
> * CPU 1 ups the refcount, seeing zero so it "does the right thing"
> 
> * CPU 2  ups the refcount,seeing the Internet from CPU 2, and treats the object as live when in fact it's being freed by CPU 1

That is IMO an invalid scenario. If CPU 0 drops refcount to 0, it means it had refcount 1 before that and it was the last owner of the object.  All other threads would have to stop using it already before that.

**Comment by kroeckx** [comment]
If there is 1 reference, one thread might want to increase and an other decrease it at the same time.

**Comment by mattcaswell** [comment]
> If there is 1 reference, one thread might want to increase and an other decrease it at the same time.

That doesn't sound like a valid scenario? This says that two threads are both sharing a single reference at the start and one thread might decrement it at any time making the shared reference invalid. Which sounds like a usage bug.

**Comment by mattcaswell** [comment]
Basically if you are going to decrement a reference then you must ensure that you are the last user of that reference.

**Comment by t8m** [comment]
> If there is 1 reference, one thread might want to increase and an other decrease it at the same time.

Think about it, it is an invalid use. If concurrently working with one object from 2 threads, where the threads are doing both up-ref and down-ref, there must be at least one reference more than 1 when a thread is up-refing. That is the reference the thread owns. So with 1 reference there cannot really be 2 threads fighting over the object doing up-ref and down-ref concurrently.

However TSAN apparently cannot cope correctly with this as the testcase is failing under TSAN.

```
==================
WARNING: ThreadSanitizer: data race (pid=23020)
  Write of size 8 at 0x7b2800005030 by thread T48:
    #0 free <null> (threadstest+0xbf424) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #1 CRYPTO_free <null> (threadstest+0x1a39b4) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #2 EVP_PKEY_free <null> (threadstest+0x18e66d) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #3 thread_release_shared_pkey threadstest.c (threadstest+0x144779) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #4 thread_run threadstest.c (threadstest+0x141d4f) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)

  Previous atomic write of size 4 at 0x7b2800005030 by thread T39:
    #0 EVP_PKEY_free <null> (threadstest+0x18e602) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #1 thread_release_shared_pkey threadstest.c (threadstest+0x144779) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #2 thread_run threadstest.c (threadstest+0x141d4f) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)

  Thread T48 (tid=23076, running) created by main thread at:
    #0 pthread_create <null> (threadstest+0xc05ed) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #1 test_multi_shared_pkey_release threadstest.c (threadstest+0x140f5d) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #2 run_tests <null> (threadstest+0x148379) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #3 main <null> (threadstest+0x148bfb) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)

  Thread T39 (tid=23067, finished) created by main thread at:
    #0 pthread_create <null> (threadstest+0xc05ed) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #1 test_multi_shared_pkey_release threadstest.c (threadstest+0x140c0f) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #2 run_tests <null> (threadstest+0x148379) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)
    #3 main <null> (threadstest+0x148bfb) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3)

SUMMARY: ThreadSanitizer: data race (/home/runner/work/openssl/openssl/test/threadstest+0xbf424) (BuildId: 575c30edaf95777d60fedabf6ba3b9c6ace9b2d3) in free
==================
```

However! Please note that this is a false positive - yes, free() writes into the object's refcount non-atomically where previous writes were atomic, but that does NOT matter because all other threads must have already stopped working with that object - otherwise they would be doing UAF.

**Comment by mattcaswell** [comment]
Is there some way to omit the free call from the TSAN checks?

**Comment by kroeckx** [comment]
I think you're saying even the acquire isn't needed

**Comment by t8m** [comment]
> I think you're saying the acquire isn't needed either

It is not completely clear to me why the acquire fence was added there.

**Comment by kroeckx** [comment]
So the reason for the acquire is to make sure all the writes from the other thread before the decrease are visible to this thread. The way to make them visible is to have one thread do a release and an other an acquire. But we currently don't actually do any release, so the acquire doesn't do anything useful. 

Without the synchronisation, the first thread to delete might have pending writes that are now possibly going to freed memory and so you'll end up with undefined behavior.

File: include/internal/refcount.h
Comment (kroeckx) [inline include/internal/refcount.h:59]: There is no reason for acquire+release, we want just release. And we want it for all versions, not just a tsan build.

**Review by kroeckx — CHANGES_REQUESTED** [review]

**Comment by t8m** [comment]
> Without the synchronisation, the first thread to delete might have pending writes that are now possibly going to freed memory and so you'll end up with undefined behavior.

So that was about the complex structures that needs some orchestration to properly work with and release anyway. Isn't it?

IMO we do not do anything like this within our refcounted objects.

File: include/internal/refcount.h
Comment (t8m) [reply include/internal/refcount.h:59]: I think @dot-asm arguments in #6900 are sound.

The acq+rel is apparently necessary for TSAN to be silent according to #25660

**Comment by t8m** [comment]
@kroeckx please feel free to submit an alternative PR. I will not fight for this one. But I will not submit a PR that I do not agree with.

**Comment by nhorman** [comment]
I think, for the sake of posterity, from what I'm reading the refcount api in openssl is essentially different from other refcount apis.  The one I am most familiar with is the refcount_t api in the linux kernel.  That api uses atomic compare and exchange in a for(;;) loop for increments/decrements.  This allows for the creation of a refcount_inc_not_zero call, which in turn allows for the use of parallel decrements/increments from different threads safely (i.e. kernel refcounting can be used locklessly when sharing a single object between two threads).

Contrast  this to the openssl refcount api, which just uses an atomic inc/dec implementation.  While thats ok, it _disallows_ the implementation of the corner use case of multiple threads doing increments/decrements between 0 and 1 (the invalid use case @mattcaswell and @t8m referred to above).  That is to say, unlike the kernels refcount mechanism, we can _only_ use refcounting locklessly here if a refcount is taken before an object is shared with another thread (i.e. thread a ups the refcount of an object, then shares it with thread b).  Barring that ordering of operations, any refcount manipulation without that prior increment must be done under the protection of an additional lock to prevent parallel increments and decrements at the 1->0 transition.

File: include/internal/refcount.h
Comment (kroeckx) [reply include/internal/refcount.h:59]: Can you at least add some comment about tsan requiring that?

**Comment by mattcaswell** [comment]
>While thats ok, it disallows the implementation of the corner use case of multiple threads doing increments/decrements between 0 and 1

I'm struggling to wrap my head around the concept that this would ever be allowed in some implementation? I guess I've not used other ref-counting apis enough?? How can it ever be valid to decrement a ref-count when you are not the last user of that reference?

**Comment by kroeckx** [comment]
So I think we need to document the assumptions for the proper use of the reference counter. People have already tried to abuse it before.

I think the main thing is that writing to the object that is reference counted requires taking a lock. We depend on the release semantics of that lock, and so don't need release semantics in the decrement.

So I currently don't see why we need the acquire. I don't see any reads that could move before the decrement that could cause problems.

**Comment by kroeckx** [comment]
So the acquire is needed in case an other thread wrote to it, and released a lock. We need to wait for all the writes to have happened before the free call.

**Comment by kroeckx** [comment]
So I think there are 2 issues for TSAN:
- It reports a false positive, it's a theoretic issue, but we shouldn't actually end up in that situation
- TSAN doesn't support atomic_thread_fence() below for the case where we take a lock and release it. We can't acquire the address of the lock, so we do a more expensive thread fence.

This change will result in TSAN not reporting the first case.

Does it also have effect on the second case? I assume it will trigger an error on one of the other addresses.

**Comment by t8m** [comment]
@nhorman IMO this is the correct fix for the issues. I've cherry-picked the REF_PRINT_COUNT() fix from your PR.

File: include/internal/refcount.h
Comment (MBkkt) [reply include/internal/refcount.h:158]: In general this PR exactly that I was expected, thanks!

Comment (t8m) [reply include/internal/refcount.h:158]: OK, replaced

**Comment by bernd-edlinger** [comment]
> > Without the synchronisation, the first thread to delete might have pending writes that are now possibly going to freed memory and so you'll end up with undefined behavior.
> 
> So that was about the complex structures that needs some orchestration to properly work with and release anyway. Wasn't it?
> 
> IMO we do not do anything like this within our refcounted objects.

i think without that aquire the following read could be scheduled before the atomic decrement,
and we would free the wrong sub-object.

**Comment by MBkkt** [comment]
@t8m I think load used to compare with one, not zero. To understand that we're last **unique** user.

For an example if some object shared between threads it cannot be write-accessed without any synchronization.
But if it's not shared it can.

So we need release-acquire sequence between being shared (make some reads in other threads) and being unique owner (make some writes)

For an example here, we probably need acquire when nref == 1
 
https://github.com/openssl/openssl/blob/36254fda37fe169e136079404a3c32aeea35cbd4/ssl/quic/quic_impl.c#L3151


Similar in other places

**Comment by bernd-edlinger** [comment]
> > i think without that aquire the following read could be scheduled before the atomic decrement,
> > and we would free the wrong sub-object.
> 
> But there is an acquire when ref count is 0 even in the current code.

Yes, I just saw the previous comments from kurt and you, where I maybe misunderstood that
it was about whether the acquire is still needed here.  I just want to confirm that it seems to
be necessary.

**Comment by t8m** [comment]
> For an example here, we probably need acquire when nref == 1

That code is run inside a mutex so IMO not relevant.

**Comment by MBkkt** [comment]
> > For an example here, we probably need acquire when nref == 1
> 
> That code is run inside a mutex so IMO not relevant.

but it's different probably unrelated to this object lock

**Comment by t8m** [comment]
> but it's different probably unrelated to this object lock

No, it is not. The mutex is global across all the stream objects in a connection.

**Comment by t8m** [comment]
ping @openssl/committers for approvals

**Review by nhorman — DISMISSED** [review]

File: include/internal/refcount.h
Comment (kroeckx) [inline include/internal/refcount.h:159]: Do you want to use  here, so it's the same as for the other platforms?

Comment (t8m) [reply include/internal/refcount.h:159]: Changed

**Comment by t8m** [comment]
ping again @nhorman @kroeckx

**Review by nhorman — APPROVED** [review]

**Review by levitte — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master, 3.4, 3.3 and 3.2 branches. Thank you for the reviews.

**Pull request closed without merging**
